catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/FormulaEditorEditTextTest.java

```diff
@@ -0,0 +1,259 @@
+
+		onBrickAtPosition(1)
+				.checkShowsText("0 ");
+	}
+
+	@Category({Cat.CatrobatLanguage.class, Level.Smoke.class})
+	@Test
+	public void testDiscardDialogDiscardSaveYes() {
+		onFormulaEditor()
+				.performEnterFormula("99");
+		pressBack();
+		onView(withText(R.string.formula_editor_discard_changes_dialog_title))
+				.check(matches(isDisplayed()));
+		pressBack();
+
+		onFormulaEditor()
+				.performEnterFormula(".9");
+		pressBack();
+		onView(withText(R.string.formula_editor_discard_changes_dialog_title))
+				.check(matches(isDisplayed()));
+		onView(withText(yes))
+				.perform(click());
+		checkToastMessage(R.string.formula_editor_changes_saved);
+		onBrickAtPosition(1)
+				.checkShowsText("99" + decimalMark + "9 ");
+	}
+
+	@Category({Cat.CatrobatLanguage.class, Level.Smoke.class})
+	@Test
+	public void testDiscardDialogDiscardSaveNo() {
+		onFormulaEditor()
+				.performEnterFormula("99.9");
+		onFormulaEditor()
+				.performClickOn(BACKSPACE);
+		onFormulaEditor()
+				.performClickOn(BACKSPACE);
+		pressBack();
+		onView(withText(R.string.formula_editor_discard_changes_dialog_title))
+				.check(matches(isDisplayed()));
+		onView(withText(no))
+				.perform(click());
+		checkToastMessage(R.string.formula_editor_changes_discarded);
+		onBrickAtPosition(1)
+				.checkShowsText("0 ");
+	}
+
+	@Category({Cat.CatrobatLanguage.class, Level.Smoke.class})
+	@Test
+	public void testFormulaIsNotValidToast() {
+		onFormulaEditor()
+				.performClickOn(BACKSPACE);
+		onFormulaEditor()
+				.performClickOn(OK);
+		checkToastMessage(R.string.formula_editor_parse_fail);
+
+		onFormulaEditor()
+				.performClickOn(BACKSPACE);
+		onFormulaEditor()
+				.performEnterFormula("+");
+		onFormulaEditor()
+				.performClickOn(OK);
+		checkToastMessage(R.string.formula_editor_parse_fail);
+
+		onFormulaEditor()
+				.performClickOn(BACKSPACE);
+		onFormulaEditor()
+				.performEnterFormula("1+1+");
+		pressBack();
+		onView(withText(R.string.formula_editor_discard_changes_dialog_title))
+				.check(matches(isDisplayed()));
+		onView(withText(yes))
+				.perform(click());
+		checkToastMessage(R.string.formula_editor_parse_fail);
+	}
+
+	@Category({Cat.CatrobatLanguage.class, Level.Smoke.class})
+	@Test
+	public void testComputeDialog() {
+		onFormulaEditor()
+				.performEnterFormula("-2");
+		onFormulaEditor()
+				.performClickOn(COMPUTE);
+		onView(withId(R.id.formula_editor_compute_dialog_textview))
+				.check(matches(withText("-2.0")));
+		pressBack();
+
+		onFormulaEditor()
+				.performEnterFormula("-6.111-");
+		onFormulaEditor()
+				.performClickOn(COMPUTE);
+		onFormulaEditor()
+				.performClickOn(BACKSPACE);
+		onFormulaEditor()
+				.performClickOn(COMPUTE);
+		onView(withId(R.id.formula_editor_compute_dialog_textview))
+				.check(matches(withText("-8.111")));
+	}
+
+	// https://stackoverflow.com/questions/28390574/checking-toast-message-in-android-espresso
+	public void checkToastMessage(int messageID) {
```

Comment:
i really dislike "check..." methods, since it is never clear what it actually does.
either rename this to "assertToastMessageDisplayed", or even better, have this function return the view matcher and rename to onToastMessage:
ViewInteraction onToastMessage(ViewMatcher matcher) {
	return onView(someMatcher)
			.inRoot(withDecorView(Matchers.not(baseActivityTestRule.getActivity().getWindow()
			.getDecorView())));
}

and then use it as:
onToastMessage(withText(messageId))
	.check(matches(isDisplayed));

this would have the advantage that you can pass anything as a matcher, and since withText() works with Ids as well as strings you get both for free.